### PR TITLE
Time-dependent TPC dEdx calibration

### PR DESCRIPTION
**Pull request by badarots**
A first attempt to write the necessary code to track the MIP position variation over time.

### CODE REVIEW

File: Detectors/TPC/calibration/CMakeLists.txt
Comment (badarots) [inline Detectors/TPC/calibration/CMakeLists.txt:40]: I have to check which library I really need to add here, I think a can remove some of those, like O2::DetectorsDCS.

Comment (badarots) [reply Detectors/TPC/calibration/CMakeLists.txt:40]: Removed

File: Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h]: I prefer to put the `Calibrator` in a separate file.

File: Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h]: Default initialize, separate lines and comments:
```suggestion
  int mMinClusters{}; ///< bla
  double mMinP{}; ///< bla
  double mMaxP{}; ///< bla
```

File: Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h]: see above, deafault initialize, doxygen comments, one variable per line.

File: Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h]: doxygen comments
```suggestion
  CcdbObjectInfoVector mInfoVector; ///< vector of CCDB Infos , each element is filled with the CCDB description of the accompanying LHCPhase
  MIPVector mMIPVector;             ///< vector of LhcPhase, each element is filled in "process" when we finalize one slot (multiple can be finalized during the same "process", which is why we have a vector. Each element is to be considered the output of the device, and will go to the CCDB
```

File: Detectors/TPC/calibration/include/TPCCalibration/FastHisto.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/FastHisto.h:25]: Why is this needed?

File: Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h]: doxygen comments

File: Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h]: is this needed here?

File: Detectors/TPC/workflow/src/CalibdEdxSpec.cxx
Comment (wiechula) [inline Detectors/TPC/workflow/src/CalibdEdxSpec.cxx]:
```suggestion
    const int slot_length = ic.options().get<int>("tf-per-slot");
    const int max_delay = ic.options().get<int>("max-delay");
    const int min_entries = std::max(50, ic.options().get<int>("min-entries"));
    const double min_p = std::min(0.49, ic.options().get<double>("min-momentum"));
    const double max_p = std::max(0.51, ic.options().get<double>("max-momentum"));
    const int min_clusters = std::max(15, ic.options().get<int>("min-clusters"));
    const int nbins = std::max(50, ic.options().get<int>("nbins"));
```

File: Detectors/TPC/workflow/src/CalibdEdxSpec.cxx
Comment (wiechula) [inline Detectors/TPC/workflow/src/CalibdEdxSpec.cxx]:
```suggestion
   const auto tfcounter = o2::header::get<DataProcessingHeader*>(pc.inputs().get("tracks").header)->startTime;
```

File: Detectors/TPC/workflow/src/CalibdEdxSpec.cxx
Comment (wiechula) [inline Detectors/TPC/workflow/src/CalibdEdxSpec.cxx]: Please add an option that this is only dumped if activated in the workflow, like here https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/workflow/include/TPCWorkflow/TPCCalibPedestalSpec.h#L180

File: Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h]: please change this everywhere
```suggestion
namespace o2::tpc
{
```

File: Detectors/TPC/workflow/src/CalibdEdxSpec.cxx
Comment (wiechula) [inline Detectors/TPC/workflow/src/CalibdEdxSpec.cxx]: at some point we should put this into a track cut class. For the moment it is fine.

**Review by wiechula — CHANGES_REQUESTED** [review]

**Review by wiechula — DISMISSED** [review]
I think for the moment this is fine. Please mark ready for review. We'll continue then with further developments in another PR.

File: Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h]:
```suggestion
  dEdxHistos(const TrackCuts& cuts, unsigned int nBins);
```
below you move, which will 'return' an empty input object. Not necessarily what one would expect I think.

File: Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h]:
```suggestion
  void dumpToFile(std::string_view fileName) const;
```

File: Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h]: Perhaps it is better to set default Track cuts and steer if they should be applied by the flag `mApplyCuts` with setters and getter.

File: Detectors/TPC/calibration/src/CalibdEdx.cxx
Comment (wiechula) [inline Detectors/TPC/calibration/src/CalibdEdx.cxx]: It might be simpler to use the `TreeStream` functionality. Then you don't need to keep branches for data memebers that should be stored in the tree. See e.g. how it is done here:
https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/calibration/include/TPCCalibration/CalibLaserTracks.h#L120
https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/calibration/src/CalibLaserTracks.cxx#L82

File: Detectors/TPC/calibration/src/dEdxHistos.cxx
Comment (wiechula) [inline Detectors/TPC/calibration/src/dEdxHistos.cxx]: see comment above

File: Detectors/TPC/calibration/src/dEdxHistos.cxx
Comment (wiechula) [inline Detectors/TPC/calibration/src/dEdxHistos.cxx]: Naming convention camel case!

File: Detectors/TPC/workflow/src/MIPTrackFilterSpec.cxx
Comment (wiechula) [inline Detectors/TPC/workflow/src/MIPTrackFilterSpec.cxx:67]: cool, I like this one!

**Review by wiechula — CHANGES_REQUESTED** [review]
Thanks for the updates! It would be good if you could squash everything, since it is hard to read the diffs now after classes were renamed. You could keep 2 commits only:
1. moving of TrackCuts and all related changes
2. all MIP calibration stuff

Afterwards I'll go through carefully once more and have a look.

For the track cuts it might be nicer to have this configurable `ParameterTrackCuts` using a `ConfigurableParam` as e.g. here: https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/base/include/TPCBase/ParameterDetector.h and add function like `setCuts(TrackCuts& cuts)`. Then one can configure the cuts easily from the command line. We can discuss details.

File: Detectors/TPC/calibration/src/CalibdEdx.cxx
Comment (badarots) [reply Detectors/TPC/calibration/src/CalibdEdx.cxx]: This is nice! I really struggled to get this data-keeping thing right, and it is so ugly. I wasn't aware of another way to do it and I am really glad you point this out. I will changes this right away!

File: Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h
Comment (badarots) [reply Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h]: You are right, this added constructor just add complexity.

File: Detectors/TPC/calibration/src/dEdxHistos.cxx
Comment (badarots) [reply Detectors/TPC/calibration/src/dEdxHistos.cxx]: It is stronger than me!

**Comment by badarots** [comment]
I reworked the commit contents as you asked and applied all requested changes, but the `ConfigurableParam` stuff.

File: Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h:76]:
```suggestion
  bool getApplyCuts() const { return mApplyCuts; }
```

File: Detectors/TPC/calibration/src/CalibdEdx.cxx
Comment (wiechula) [inline Detectors/TPC/calibration/src/CalibdEdx.cxx]: perhaps `enableDebugStream` or `enableDebugOutput`.

File: Detectors/TPC/calibration/src/CalibdEdx.cxx
Comment (wiechula) [inline Detectors/TPC/calibration/src/CalibdEdx.cxx]: Same as above.

File: Detectors/TPC/calibration/src/CalibdEdx.cxx
Comment (wiechula) [inline Detectors/TPC/calibration/src/CalibdEdx.cxx]: Also here the name is a bit misleading, since the dumping to file using the stream functionality is done in the place where it is used. Usually, dumpToFile should really put contents to a file. But this is not what is done here. So one might call it `finalizeDebugOutput` or so.

File: Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h]: perhaps
```suggestion
  const std::array<Hist, 2>& getHists() const { return mHist; }
```
since it is not only one.

**Review by wiechula — CHANGES_REQUESTED** [review]
Thanks a lot @badarots! Only minor things left. Please amend them to the latest commit.

**Comment by badarots** [comment]
Hi @wiechula, I applied the requested changes! As you can see I changed the `CalibMIP` from a `std::array` to a custom struct, and added it to the root dictionary. Now the debug output to file is working correctly, but this needs to be changed in the future.

**Review by wiechula — DISMISSED** [review]
@badarots , thanks for all the changes.

**Comment by badarots** [comment]
What now @wiechula? Should I squash the last two commits and proceed to the merge of this PR or do you prefer that I finish the `ParameterTrackCuts` first?

**Comment by wiechula** [comment]
@badarots , merging can only be done by the code owners. Yes, please squash the last two commits. I'll then approve once more and after all test have finished successfully I'll ask to have it merged.

**Comment by wiechula** [comment]
And make it ready for review, of course.

**Review by wiechula — APPROVED** [review]

**Review by shahor02 — CHANGES_REQUESTED** [review]
@badarots the o2 CI failure is genuine:
````
/mnt/mesos/sandbox/sandbox/o2/sw/SOURCES/O2/6305/0/Detectors/TPC/calibration/src/dEdxHistos.cxx: In member function 'void o2::tpc::dEdxHistos::merge(const o2::tpc::dEdxHistos*)':
/mnt/mesos/sandbox/sandbox/o2/sw/SOURCES/O2/6305/0/Detectors/TPC/calibration/src/dEdxHistos.cxx:57:24: error: 'prev' was not declared in this scope; did you mean 'std::prev'?
   57 |     assert(binCount == prev->getHist()[i].getNBins());
      |                        ^~~~
````

File: Detectors/TPC/calibration/src/dEdxHistos.cxx
Comment (wiechula) [inline Detectors/TPC/calibration/src/dEdxHistos.cxx]: I guess this should be
```suggestion
    assert(binCount == other->getHist()[i].getNBins());
```
@badarots , can you please check.

**Review by wiechula — CHANGES_REQUESTED** [review]

**Comment by badarots** [comment]
As discussed with @wiechula today, we agreed to just remove that assertion line.

**Comment by badarots** [comment]
So there is an error in the `ensure-executable-naming-convention` test
```
ERROR: executable name o2-tpc-calib-dEdx does not follow naming convention : should only contains lowercase letters, numbers, or dashes
```

Should I just change the executable name to `o2-tpc-calib-dedx` (lowercase "de")?

File: Detectors/TPC/workflow/CMakeLists.txt
Comment (shahor02) [inline Detectors/TPC/workflow/CMakeLists.txt]: yes, just change this to `calib-dedx`

**Comment by wiechula** [comment]
@shahor02 , @davidrohr , I think the macOS failure is unrelated.

**Comment by davidrohr** [comment]
Probably, I have seen such errors often on the MacCI recently.
@TimoWilken @MichaelLettrich : Could you check why the MacCI fails here?

**Comment by shahor02** [comment]
I thought this problem of compiled macro's test was fixed already... Anyway, the macOS log also contains legitimate warnings:
````
/System/Volumes/Data/build/ci_checks/alice_ci_0/o2/sw/SOURCES/O2/6305/0/Detectors/TPC/qc/macro/runPID.C:55:23: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare]
    for (int k = 0; k < nTracks; k++) {
````
````
/System/Volumes/Data/build/ci_checks/alice_ci_0/o2/sw/SOURCES/O2/6305/0/Detectors/TPC/qc/macro/runTracks.C:59:23: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare]
    for (int k = 0; k < nTracks; k++) {
                    ~ ^ ~~~~~~~
````

**Comment by wiechula** [comment]
Indeed, but these lines were not touched by this commit.
@badarots , can you fix it nevertheless in both places:
```c++
for (size_t k = 0; k < nTracks; k++) {
```
should do.

**Comment by badarots** [comment]
I changed `int` -> `size_t` as requested. I'm not sure if I was supposed to amend my last commit or create a new one for this little change. So I amended it.

**Review by wiechula — APPROVED** [review]

**Comment by wiechula** [comment]
@shahor02 , can this be merged? I think the macOS failure is again unrelated. cs8 fails with `No space left on device`.

**Review by shahor02 — APPROVED** [review]

**Pull request merged**
